ext/uri/php_uri.c

```diff
@@ -40,12 +47,560 @@ static const zend_module_dep uri_deps[] = {
+{
+	ZEND_PARSE_PARAMETERS_NONE();
+
+	zend_object *this_object = Z_OBJ_P(ZEND_THIS);
+	uri_internal_t *internal_uri = uri_internal_from_obj(this_object);
+	URI_ASSERT_INITIALIZATION(internal_uri);
+
+	RETURN_STR(internal_uri->handler->uri_to_string(internal_uri->uri, URI_RECOMPOSITION_RAW_UNICODE, false));
+}
+
+PHP_METHOD(Uri_WhatWg_Url, toAsciiString)
+{
+	ZEND_PARSE_PARAMETERS_NONE();
+
+	zend_object *this_object = Z_OBJ_P(ZEND_THIS);
+	uri_internal_t *internal_uri = uri_internal_from_obj(this_object);
+	URI_ASSERT_INITIALIZATION(internal_uri);
+
+	RETURN_STR(internal_uri->handler->uri_to_string(internal_uri->uri, URI_RECOMPOSITION_RAW_ASCII, false));
+}
+
+PHP_METHOD(Uri_WhatWg_Url, resolve)
+{
+	zend_string *uri_str;
+	zval *errors = NULL;
+
+	ZEND_PARSE_PARAMETERS_START(1, 2)
+		Z_PARAM_PATH_STR(uri_str)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_ZVAL(errors)
+	ZEND_PARSE_PARAMETERS_END();
+
+	zend_object *this_object = Z_OBJ_P(ZEND_THIS);
+	uri_internal_t *internal_uri = uri_internal_from_obj(this_object);
+	URI_ASSERT_INITIALIZATION(internal_uri);
+
+	php_uri_instantiate_uri(INTERNAL_FUNCTION_PARAM_PASSTHRU, internal_uri->handler, uri_str, this_object, true, false, errors);
+}
+
+PHP_METHOD(Uri_WhatWg_Url, __serialize)
+{
+	ZEND_PARSE_PARAMETERS_NONE();
+
+	zend_object *this_object = Z_OBJ_P(ZEND_THIS);
+	uri_internal_t *internal_uri = uri_internal_from_obj(this_object);
+	URI_ASSERT_INITIALIZATION(internal_uri);
+
+	/* Serialize state: "uri" key in the first array */
+	zend_string *uri_str = internal_uri->handler->uri_to_string(internal_uri->uri, URI_RECOMPOSITION_RAW_ASCII, false);
+	if (uri_str == NULL) {
+		zend_throw_exception_ex(NULL, 0, "Cannot recompose %s to string", ZSTR_VAL(this_object->ce->name));
+		RETURN_THROWS();
+	}
+	zval tmp;
+	ZVAL_STR(&tmp, uri_str);
+
+	array_init(return_value);
+
+	zval arr;
+	array_init(&arr);
+	zend_hash_str_add_new(Z_ARRVAL(arr), ZEND_STRL(URI_SERIALIZED_PROPERTY_NAME), &tmp);
+	zend_hash_next_index_insert(Z_ARRVAL_P(return_value), &arr);
+
+	/* Serialize regular properties: second array */
+	ZVAL_ARR(&arr, this_object->handlers->get_properties(this_object));
+	Z_TRY_ADDREF(arr);
+	zend_hash_next_index_insert(Z_ARRVAL_P(return_value), &arr);
+}
+
+PHP_METHOD(Uri_WhatWg_Url, __unserialize)
+{
+	uri_unserialize(INTERNAL_FUNCTION_PARAM_PASSTHRU, URI_PARSER_WHATWG);
+}
+
+PHP_METHOD(Uri_WhatWg_Url, __debugInfo)
+{
+	ZEND_PARSE_PARAMETERS_NONE();
+
+	zend_object *object = Z_OBJ_P(ZEND_THIS);
+
+	RETURN_ARR(uri_get_debug_properties(object));
+}
+
+static zend_object *uri_create_object_handler(zend_class_entry *class_type)
+{
+	uri_object_t *uri_object = zend_object_alloc(sizeof(uri_object_t), class_type);
+
+	zend_object_std_init(&uri_object->std, class_type);
+	object_properties_init(&uri_object->std, class_type);
+
+	return &uri_object->std;
+}
+
+static void uri_free_obj_handler(zend_object *object)
+{
+	uri_object_t *uri_object = uri_object_from_obj(object);
+	ZEND_ASSERT(uri_object != NULL);
```

Comment:
This assertion is pretty useless, since `uri_object_from_obj()` just does some pointer arithmetic:

```suggestion
```